### PR TITLE
Added one more way to send emails - gmail via smtp, added interface for...

### PR DESCRIPTION
**Pull request by vantihovich**
… switching between mailjet and smtp

### CODE REVIEW

File: swagger/main.go [hunk 1]
```diff
@@ -106,10 +117,21 @@ func service() http.Handler {
...
+	UsersProvider := handlers.UsersHandler{}
+
+	if cfgEmail.MailChannel == "mailjet" {
+		UsersProvider = *handlers.NewUsersHandler(&db, cache, cfgJWT.SecretKey, cfgLogin, clientMailJet)
+	} else if cfgEmail.MailChannel == "gmail" {
+		UsersProvider = *handlers.NewUsersHandler(&db, cache, cfgJWT.SecretKey, cfgLogin, clientGmailSmtp)
+	}
 
-	UsersProvider := handlers.NewUsersHandler(&db, cache, cfgJWT.SecretKey, cfgLogin, mailClient)
+	//UsersProvider := handlers.NewUsersHandler(&db, cache, cfgJWT.SecretKey, cfgLogin, mailingClient)
```
Comment (apaliavy) [inline swagger/main.go:134]: commented code

File: swagger/configuration/configs.go [hunk 2]
```diff
@@ -33,11 +33,22 @@ type MailJetParameters struct {
...
+	Port     string `env:"SMTPPORT,required"`
+}
+
 type WorldCoinIndexParameters struct {
 	Key string `env:"WORLDCOININDEXKEY,required"`
 	URL string `env:"WORLDCOININDEXURL,required"`
 }
 
+type EmailParameters struct {
+	MailChannel string `env:"MAILCHANNEL,required"`
```
Comment (apaliavy) [inline swagger/configuration/configs.go:49]: MailProvider

File: swagger/handlers/users.go [hunk 3]
```diff
@@ -406,13 +405,28 @@ func (h *UsersHandler) ForgotPasswordSendEmail(w http.ResponseWriter, r *http.Re
...
-	err = h.mailClient.SendForgottenPasswordEmail(parameters.Email, secret)
+
+	err = h.emailSender.SendForgottenPasswordEmail(parameters.Email, secret)
 	if err != nil {
 		log.WithError(err).Info("An attempt to send email to user`s email returned error")
 		w.WriteHeader(http.StatusInternalServerError)
 		return
 	}
 
+	// err = h.mailClient.SendForgottenPasswordEmail(parameters.Email, secret)
```
Comment (apaliavy) [inline swagger/handlers/users.go:416]: ?

Comment (vantihovich) [reply swagger/handlers/users.go:416]: I thought, that I`ve deleted all that, smth went wrong, sorry((

File: swagger/configuration/configs.go (hunk 2 above)
Comment (vantihovich) [reply swagger/configuration/configs.go:49]: done

File: swagger/main.go (hunk 1 above)
Comment (vantihovich) [reply swagger/main.go:134]: removed

File: swagger/main.go [hunk 4]
```diff
@@ -106,10 +117,20 @@ func service() http.Handler {
...
-	mailClient := email.New(cfgMailJet)
+	log.Info("connecting to MailJet service")
+	clientMailJet := mailjet.New(cfgMailJet)
+
+	log.Info("authenticating to GmailSmtp service")
+	clientGmailSmtp := gmailsmtp.Auth(cfgSmtpGmail)
+
+	UsersProvider := handlers.UsersHandler{}
+
+	if cfgEmail.MailProvider == "mailjet" {
```
Comment (apaliavy) [inline swagger/main.go:128]: What the hell is happening here? Please explain your logic 😱

Comment (vantihovich) [reply swagger/main.go:128]: There is an env parameter called MAILCHANNEL,  the values for this parameter might be "mailjet" or "gmail", so depending on this parameter the the appropriate mailing provider will be used for sending emails. This logic is realized via Interface.

Comment (vantihovich) [reply swagger/main.go:128]: fixed

File: swagger/main.go [hunk 5]
```diff
@@ -9,14 +9,17 @@ import (
 	"syscall"
 	"time"
 
+	"github.com/vantihovich/go_tasks/tree/master/swagger/models"
```
Comment (apaliavy) [inline swagger/main.go:12]: imports are not sorted

File: swagger/gmailsmtp/smtp.go [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+	authenticate := smtp.PlainAuth("", cfg.Login, cfg.Password, cfg.Host)
+	return SmtpClient{
+		Client: authenticate,
+		Host:   cfg.Host,
+		Port:   cfg.Port,
+		Sender: cfg.Login,
+	}
+}
+
+const tmpl = `<h3>Dear logginer, here is your <a href="http://localhost:3000/auth/forgot_password_reset_password?parameter={{.Secret}}">link</a> for changing the password in our services.</h3><br />May the force be with you!`
```
Comment (apaliavy) [inline swagger/gmailsmtp/smtp.go:30]: can you please move it to the beginning of the file

File: swagger/gmailsmtp/smtp.go [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+		Secret: secret,
+	}
+
+	buf := new(bytes.Buffer)
+	if err = parsedTemplate.Execute(buf, data); err != nil {
+		log.WithError(err).Info("Error writing the message to buffer after applying the data to the template for smtp")
+		return err
+	}
+
+	msg := []byte("To:" + receiver + "\r\n" +
```
Comment (apaliavy) [inline swagger/gmailsmtp/smtp.go:55]: can you please build the string using fmt.Sprintf()

File: swagger/main.go [hunk 8]
```diff
@@ -106,10 +119,16 @@ func service() http.Handler {
...
+	if cfgEmail.MailProvider == "mailjet" {
+		log.Info("connecting to MailJet service")
+		mailProvider = mailjet.New(cfgMailJet)
+	} else if cfgEmail.MailProvider == "gmail" {
+		log.Info("authenticating to GmailSmtp service")
+		mailProvider = gmailsmtp.Auth(cfgSmtpGmail)
+	}
 
-	UsersProvider := handlers.NewUsersHandler(&db, cache, cfgJWT.SecretKey, cfgLogin, mailClient)
+	UsersProvider := handlers.NewUsersHandler(&db, cache, cfgJWT.SecretKey, cfgLogin, mailProvider)
```
Comment (apaliavy) [inline swagger/main.go:131]: Quick question - why UsersProvider is exported? Do you use it outside of this package?

File: swagger/main.go (hunk 5 above)
Comment (vantihovich) [reply swagger/main.go:12]: fixed

File: swagger/gmailsmtp/smtp.go (hunk 6 above)
Comment (vantihovich) [reply swagger/gmailsmtp/smtp.go:30]: done

File: swagger/main.go (hunk 8 above)
Comment (vantihovich) [reply swagger/main.go:131]: nope, made usersProvider and wciProvider not exported.

File: swagger/gmailsmtp/smtp.go (hunk 7 above)
Comment (vantihovich) [reply swagger/gmailsmtp/smtp.go:55]: done

**Review by apaliavy — APPROVED** [review]

**Pull request merged**